### PR TITLE
chore: add client extensions examples to existing Prisma Client sections

### LINKED ISSUE #4481: Link to PCE blog post from the docs
There should be several links from the PCE docs pages to the [blog post](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn):

- Root (https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions): Add a note to the intro with a link to the blog post, mentioning that it contains a lot of practical examples
- Go through the various docs pages for PCE and find relevant places where to link specific examples of the blog post

### PR DESCRIPTION
**Pull request by ruheni**
## Describe this PR

Closes #4481 

Add Prisma Client extension examples to existing workarounds:
Show an example with Prisma Client Extensions on the page

- [x] Computed fields
- [x] Custom validation
- [x] Custom models methods

Removed from the PR
- [ ] ~~`exists` function~~
- [x] ~~JSON field type validation~~
- [x] ~~Middleware migration to Client extensions: ~~
	- [ ] ~~soft delete~~ (Update when support for nesting is added)
	- [x] ~~query logging~~
	- [ ] ~~session data~~ (RLS?)
- [ ] ~~RLS~~
- [x] ~~Move PCE a level above Middleware?~~
- [x] ~~Set up redirects for old pages~~

Examples marked with `~~` are currently out of scope for v1.

**Alternative idea**:
Convert the [Prisma Client extensions GH repo](https://github.com/prisma/prisma-client-extensions) to resemble the Prisma Examples repo — in terms of structure and purpose.
The repository can serve as a reference + dump for PCE ideas + contributions from other users.
Create a page to aggregate the list of all extensions in the repository — name and descriptions

**Note**: Closed in favor of #4719

### CODE REVIEW

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return result !== null
+      },
+    },
+  },
+})
+
+// Example queries
+await prisma.post.exists(1)
+
+await prisma.user.exists(1)
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx]:
```suggestion
  await prisma.user.exists({
    name: 'Alice'
  })

  await prisma.post.exists({
    OR: [
      { title: { contains: 'Prisma' } },
      {content: {contains: 'Prisma'}}
    ]
  })
```

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+// Example queries
+  await prisma.user.exists({
+    name: 'Alice'
+  })
+
+  await prisma.post.exists({
+    OR: [
+      { title: { contains: 'Prisma' } },
+      {content: {contains: 'Prisma'}}
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx]:
```suggestion
      { content: { contains: 'Prisma' } }
```

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  await prisma.user.exists({
+    name: 'Alice'
+  })
+
+  await prisma.post.exists({
+    OR: [
+      { title: { contains: 'Prisma' } },
+      { content: { contains: 'Prisma' } }
+    ]
+  })
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/202-exists-function.mdx]:
```suggestion
await prisma.user.exists({
  name: 'Alice'
})

await prisma.post.exists({
  OR: [
    { title: { contains: 'Prisma' } },
    { content: { contains: 'Prisma' } }
  ]
})
```

File: content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx [hunk 4]
```diff
@@ -51,36 +51,31 @@ generator client {
...
+  id    Int     @id @default(autoincrement())
+  name  String?
+  email String  @unique
 
-enum Role {
-  ADMIN
-  USER
-  MODERATOR
+  posts Post[]
 }
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx]: Why this change to the logging middleware?

**Comment by janpio** [comment]
What is the difference between an "example" and a "sample"?
image

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/201-custom-validation.mdx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+- Learn more sample use cases you could use Prisma Client extensions for in [this blog post](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn).
+- You can find more Prisma Client extensions sample cases in this [GitHub repository](https://github.com/prisma/prisma-client-extensions)
+
+{/\*\*
+
+## Going Further
+
+- Learn how you can create and modify queries using [Prisma Client extensions](/concepts/components/prisma-client/client-extensions/query)
+- Learn how you can organize your code better by moving the `signup` function into [a custom model](/concepts/components/prisma-client/custom-models).
+  \*/}
```
Comment (janpio) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/201-custom-validation.mdx]: Broken output:
image

**Comment by ruheni** [comment]
> Difference between "sample" and "example"

I'm still conflicted on what terminology to stick to. The docs on Middleware uses "sample", however, it doesn't feel as natural. I'm open to suggestions and don't have any particular preferences. 

The "Extensions Examples" page lists examples listed in the [prisma-client-extensions repo](https://github.com/prisma/prisma-client-extensions), and "EXT sample" shows how to implement one. I'm also unsure whether we should provide the implementation of samples or list examples in the repo. The former might not be a great approach if we intend to add more examples as it might get too big.

File: content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx (hunk 4 above)
Comment (ruheni) [reply content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx]: The intention was to simplify the data model being used by removing "unnecessary" fields and make it consistent with the default data model we use for the docs. I would by happy to undo this change if you don't think its a necessary change.

File: vercel.json [hunk 6]
```diff
@@ -1143,6 +1143,14 @@
...
       "destination": "/docs/data-platform/troubleshooting/cannot-change-db-of-env"
     },
+    {
+      "source": "/docs/concepts/components/prisma-client/custom-validation",
+      "destination": "docs/concepts/components/prisma-client/client-extensions/custom-validation"
+    },
+    {
+      "source": "/docs/concepts/components/prisma-client/computed-fields",
+      "destination": "docs/concepts/components/prisma-client/client-extensions/computed-fields"
+    },
```
Comment (janpio) [inline vercel.json]: Unfortunately redirects are not testable @nilubava.

This is the preview URL that is rendered in the Vercel links comment: https://docs-git-client-extensions-prisma.vercel.app/concepts/components/prisma-client/custom-validation This is a 404.

If I use the `source` URL, I get https://docs-git-client-extensions-prisma.vercel.app/docs/concepts/components/prisma-client/custom-validation - which redirects to a broken URL which is also 404.

**Comment by janpio** [comment]
We do not consider "Custom validation" and "Computed fields" solved, so their feature request issues are still open. They ask for a top level feature, not the building blocks and then a "sample/example" that is marked as "do not use in production". https://www.prisma.io/docs/concepts/components/prisma-client/custom-validation and https://www.prisma.io/docs/concepts/components/prisma-client/computed-fields were clear about that, https://docs-git-client-extensions-prisma.vercel.app/concepts/components/prisma-client/client-extensions/custom-validation and https://docs-git-client-extensions-prisma.vercel.app/concepts/components/prisma-client/client-extensions/computed-fields are not.

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/199-extension-examples.mdx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+| [`readonly-client`](https://github.com/prisma/prisma-client-extensions/readonly-client)           | Creates a client that only allows read operations                                           |
+| [`retry-transactions`](https://github.com/prisma/prisma-client-extensions/retry-transactions)     | Adds a retry mechanism to transactions with exponential backoff and jitter                  |
+| [`row-level-security`](https://github.com/prisma/prisma-client-extensions/row-level-security)     | Uses Postgres row-level security policies to isolate data a multi-tenant application        |
+| [`static-methods`](https://github.com/prisma/prisma-client-extensions/static-methods)             | Adds custom query methods to Prisma Client models                                           |
+| [`transformed-fields`](https://github.com/prisma/prisma-client-extensions/transformed-fields)     | Demonstrates how to use result extensions to transform query results and add i18n to an app |
+| [`exists-method`](https://github.com/prisma/prisma-client-extensions/exists-method)               | Demonstrates how to add an `exists` method to all your models                               |
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/199-extension-examples.mdx]: these all 404, it should be e.g. `https://github.com/prisma/prisma-client-extensions/tree/main/audit-log-context` instead of `https://github.com/prisma/prisma-client-extensions/audit-log-context`

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/199-extension-examples.mdx [hunk 8]
```diff
@@ -0,0 +1,37 @@
+---
+title: 'Extension examples'
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/053-client-extensions/199-extension-examples.mdx]: Why not just call it "Examples"? 

"Extension" is redundant here because we're already in the "Prisma Client Examples" section.

File: content/200-concepts/100-components/02-prisma-client/053-client-extensions/201-custom-validation.mdx (hunk 5 above)
Comment (nikolasburk) [reply content/200-concepts/100-components/02-prisma-client/053-client-extensions/201-custom-validation.mdx]: Came here to note the same 🙃

**Review by nikolasburk — CHANGES_REQUESTED** [review]
It's not super clear to me why we include the sub-pages with the examples. Aren't they just the same examples from the [blog post](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-json-field-types)? What's the benefit of repeating them in the docs here? 

Also about the naming in the sidenav: Not a huge fan of the "Extension sample"-prexif (same for the Middleware section actually), e.g. "Extension sample: Logging". A couple of other ideas:

- "Example: Logging" (Also think the example names should be uppercase after the colon)
- "Use case: Logging"
- "Logging"

**Comment by ruheni** [comment]
> It's not super clear to me why we include the sub-pages with the examples. Aren't they just the same examples from the [blog post](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-json-field-types)? 

When initially working on the content, the docs were a good place to cross-post the examples created in the blog post. We can aggregate the content in a table instead of creating a separate page for each example and duplicating what's in the blog post + PCE examples repo.

I'll also revert the changes made to the pages on middleware and have them in a separate PR.

File: content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx (hunk 4 above)
Comment (ruheni) [reply content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx]: I've restored the model used here. I'll simplify the model in a separate PR

File: content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx [hunk 9]
```diff
@@ -84,3 +84,7 @@ enum Role {
 ```
 
 </TopBlock>
+
+## Going further
+
+You can also use [Prisma Client extensions](/concepts/components/prisma-client/client-extensions) to log the time take to perform a query. You can see the example here: [Extension sample: logging](/concepts/components/prisma-client/client-extensions/logging)
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/054-middleware/200-logging-middleware.mdx]:
```suggestion
You can also use [Prisma Client extensions](/concepts/components/prisma-client/client-extensions) to log the time it takes to perform a query. You can also find a functional example in [this GitHub repository](https://github.com/prisma/prisma-client-extensions/tree/main/query-logging).
```

File: content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx [hunk 10]
```diff
@@ -89,3 +89,7 @@ async function main() {
 ```
 
 Now you can use your custom `signup` method alongside `count`, `updateMany`, `groupBy` and all of the other wonderful methods that Prisma Client provides. Best of all, it's all type-safe!
+
+## Going further
+
+We recommend using [Prisma Client extensions](/concepts/components/prisma-client/client-extensions) to extend your models with [custom model methods](/concepts/components/prisma-client/client-extensions/custom-model-methods).
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/064-custom-models.mdx]:
```suggestion
We recommend using [Prisma Client extensions](/concepts/components/prisma-client/client-extensions) to extend your models with [custom model methods](https://github.com/prisma/prisma-client-extensions/tree/main/instance-methods).
```

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 11]
```diff
@@ -19,10 +19,10 @@ Here's an example using [Superstruct](https://github.com/ianstormtaylor/superstr
 ```tsx
 import { PrismaClient, Prisma, User } from '@prisma/client'
 import { assert, object, string, size, refine } from 'superstruct'
+
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx]:
```suggestion
```

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 12]
```diff
@@ -22,7 +22,6 @@ import { assert, object, string, size, refine } from 'superstruct'
 import isEmail from 'isemail'
 
 const prisma = new PrismaClient()
-
 // Runtime validation
```
Comment (ruheni) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx]:
```suggestion

// Runtime validation
```

**Comment by ruheni** [comment]
I've made a few updates to the PR:
- Restored the custom validation and computed fields file. I've linked to Prisma Client extensions in the "Going further" section
- Restored the redirects file (no redirects needed as I've restored the file)
- Deleted all files with extension examples. We're only listing and linking to the Prisma Client extensions repo examples
- Restored the model used in the logging middleware example (will be taken care of in a separate PR)

**Comment by janpio** [comment]
```
/vercel/path0/content/800-data-platform/400-members/100-add.mdx: Expected corresponding JSX closing tag for <Admonition>. (34:0)
  32 | <p parentName="li">{`From the `}<strong parentName="p">{`Role`}</strong>{` drop-down menu, select a `}<a parentName="p" {...{"href":"/data-platform/members/roles-permissions"}}>{`role`}</a>{` for the new team member.`}</p>
  33 | <Admonition>
> 34 | </li>
     | ^
  35 | </ol>
  36 | <p><strong parentName="p">{`Note`}</strong><br /><br /></p>
  37 | <p>{`On the `}<a parentName="p" {...{"href":"/data-platform/billing/plans-and-quotas#strongfreestrong-plan"}}>{`Free plan`}</a>{`, you can only assign the `}<a parentName="p" {...{"href":"/data-platform/members/roles-permissions"}}>{`Admin`}</a>{` role.`}<br /><br />{`
``` 
@ruheni

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 13]
```diff
@@ -6,12 +6,119 @@ metaDescription: 'This page explains how to add custom validation to Prisma Clie
...
+    content: 'Post content',
+  },
+})
+```
+
+</tab>
+
+<tab>
+
+```prisma file="prisma/schema.prisma" copy
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx]: the extra file name looks a bit glitchy when used in combination with tabs: 

image


```suggestion
```prisma copy
```

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 14]
```diff
@@ -6,12 +6,119 @@ metaDescription: 'This page explains how to add custom validation to Prisma Clie
...
+  authorId Int?
+  author   User? @relation(fields: [authorId], references: [id])
+}
+```
+
+</tab>
+
+</TabbedContent>
+
 ## Custom Signup Validation
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx]:
```suggestion
## Custom signup validation
```

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 15]
```diff
@@ -52,5 +159,6 @@ The example above shows how you can create a custom type-safe `signup` function
 
 ## Going Further
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx]:
```suggestion
## Going further
```

File: content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx [hunk 16]
```diff
@@ -6,12 +6,119 @@ metaDescription: 'This page explains how to add custom validation to Prisma Clie
...
+  }
+})
+```
+
+</tab>
+
+<tab>
+
+```ts copy
+// A post with invalid input — missing the title field
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx:76]: it's nice that you're showing an example query here, but I don't find it too helpful without further context and hidden in the tabs. how is it going to help someone reading this page understand validation better?

it would be more helpful IMO if:
- the query wasn't hidden in a tab
- there were two scenarios shown: one with a valid input one another one with an invalid input, including an explanation of the expected error

File: content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx [hunk 17]
```diff
@@ -10,6 +10,81 @@ Computed fields allow you to derive a new field based on existing data. A common
...
+  },
+})
+```
+
+</tab>
+
+<tab>
+
+```ts
+const user = await prisma.user.findMany({
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx:45]: it's nice that you're showing an example query here, but I don't find it too helpful without further context and hidden in the tabs. how is it going to help someone reading this page understand computed fields better?

it would be more helpful IMO if:
- the query wasn't hidden in a tab
- we showed the actual resulting structure that includes the new field and a short explanation

File: content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx [hunk 18]
```diff
@@ -10,6 +10,81 @@ Computed fields allow you to derive a new field based on existing data. A common
...
+  authorId Int?
+  author   User? @relation(fields: [authorId], references: [id])
+}
+```
+
+</tab>
+
+</TabbedContent>
+
 ## Deriving a Full Name from a First and Last Name
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx:88]: I know that this isn't the part that you've modified but when working on a docs page, let's try to look at it holistically and see what we can improve. since we use regular sentence casing for headings, these are small wins to make the docs more consistent.

File: content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx [hunk 19]
```diff
@@ -76,5 +151,6 @@ With this function, any object that contains `firstName` and `lastName` keys can
 
 ## Going Further
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/062-computed-fields.mdx]:
```suggestion
## Going further
```

**Review by nikolasburk — CHANGES_REQUESTED** [review]
looks great, I've left a few small comments but once there are incorporated, this is ready to merge 🚀
